org.jcryptool.visual.rabin/src/org/jcryptool/visual/rabin/GUIHandler.java

```diff
@@ -0,0 +1,1497 @@
+				
+				txtP.setBackground(ColorService.RED);
+				qtmp = new BigInteger(qAsStr);
+				
+				if(qtmp.compareTo(limitUp) > 0) {
+					txtQ.setBackground(ColorService.RED);
+					txtWarning.setText("Attention:\n1) only an upper limit of 2^" + limitExp + " is allowed\n" //$NON-NLS-1$ //$NON-NLS-2$
+								+ "2) p is missing"); //$NON-NLS-1$
+					showControl(txtWarning);
+					return;
+				}
+				
+				if(rabinFirst.isSuitablePrime(qtmp)) {
+					txtQ.setBackground(ColorService.GREEN);
+					txtWarning.setText("Attention: p is missing"); //$NON-NLS-1$
+					showControl(txtWarning);
+				}
+				else {
+					txtQ.setBackground(ColorService.RED);
+					txtWarning.setText("Attention:\n" //$NON-NLS-1$
+							+ "1) p is missing\n" //$NON-NLS-1$
+							+ "2) q is not a suitable prime"); //$NON-NLS-1$
+					showControl(txtWarning);
+				}
+				return;
+			}
+			
+			if(!pAsStr.isEmpty() && qAsStr.isEmpty()) {
+			
+				
+
+				if(!pAsStr.matches(pattern)) {
+					txtP.setBackground(ColorService.RED);
+					txtQ.setBackground(ColorService.WHITE);
+					txtWarning.setText(strNumberRestriction);
+					showControl(txtWarning);
+					return;
+				}
+				
+				ptmp = new BigInteger(pAsStr);
+				txtQ.setBackground(ColorService.RED);
+				
+				
+				if(ptmp.compareTo(limitUp) > 0) {
+					txtP.setBackground(ColorService.RED);
+					txtWarning.setText("Attention:\n" //$NON-NLS-1$
+						+ "1) only an upper limit of 2^" + limitExp + " is allowed\n" //$NON-NLS-1$ //$NON-NLS-2$
+								+ "2) q is missing"); //$NON-NLS-1$
+					showControl(txtWarning);
+					return;
+				}
+				
+				
+				if(rabinFirst.isSuitablePrime(ptmp)) {
+					txtP.setBackground(ColorService.GREEN);
+					txtWarning.setText("Attention: q is missing"); //$NON-NLS-1$
+					showControl(txtWarning);
+				}
+				else {
+					txtP.setBackground(ColorService.RED);
+					txtWarning.setText("Attention:\n" //$NON-NLS-1$
+							+ "1) q is missing\n" //$NON-NLS-1$
+							+ "2) p is not a suitable prime"); //$NON-NLS-1$
+					showControl(txtWarning);
+				}
+				return;
+			}
+			
+			if(!pAsStr.isEmpty() && !qAsStr.isEmpty()) {
+			
+				
+				boolean checkPattern = pAsStr.matches(pattern) && qAsStr.matches(pattern);
+				
+				if(!checkPattern) {
+					if(!pAsStr.matches(pattern) && !qAsStr.matches(pattern)) {
+						txtP.setBackground(ColorService.RED);
+						txtQ.setBackground(ColorService.RED);
+						txtWarning.setText(strNumberRestriction);
+						showControl(txtWarning);
+					}
+					if(!pAsStr.matches(pattern) && qAsStr.matches(pattern)) {
+						txtP.setBackground(ColorService.RED);
+						txtQ.setBackground(ColorService.WHITE);
+						txtWarning.setText(strNumberRestriction);
+						showControl(txtWarning);
+					}
+					if(pAsStr.matches(pattern) && !qAsStr.matches(pattern)) {
+						txtP.setBackground(ColorService.WHITE);
+						txtQ.setBackground(ColorService.RED);
+						txtWarning.setText(strNumberRestriction);
+						showControl(txtWarning);
+					}
+					return;
+					
+				}
```

Comment:
The `String.match()` method is called over and over.

It is better to declare both matches as variables and use them. Also, this can be simplified. You can see that the only thing changing is the Color. 
```suggestion
                                boolean pValid = pAsStr.matches(pattern);
                                boolean qValid = qAsStr.matches(pattern);
				boolean bothValid =  pValid
				
				if (!bothValid) {
				    var pColor = pValid ? ColorService.WHITE : ColorService.RED;
				    var qColor = qValid ? ColorService.WHITE : ColorService.RED;
				    txtWarning.setText(strNumberRestriction);
				    showControl(txtWarning);
				    return;
				}			
```